src/server/worker/branch_set.go

```diff
@@ -51,81 +58,207 @@ func (a *APIServer) newBranchSetFactory(_ctx context.Context) (branchSetFactory,
-							return false
+						for _, rootInput := range rootInputs {
+							if rootInput.Atom != nil {
+								if len(rootCommits[rootInput.Atom.Repo]) != 1 {
+									return false
+								}
+							}
+						}
+						return true
+					}); set != nil {
+						bs := &branchSet{
+							NewBranch: i,
+						}
+						for i, commitInfo := range set {
+							branch := "master"
+							if directInputs[i].Atom != nil {
+								branch = directInputs[i].Atom.Branch
+							}
+							bs.Branches = append(bs.Branches, &pfs.BranchInfo{
+								Name: branch,
+								Head: commitInfo.Commit,
+							})
+						}
+						select {
+						case <-ctx.Done():
+							commitSetsMutex.Unlock()
+							return
+						case ch <- bs:
 						}
 					}
-					return true
-				}); set != nil {
-					bs := &branchSet{
-						NewBranch: i,
-					}
-					for i, commitInfo := range set {
-						bs.Branches = append(bs.Branches, &pfs.BranchInfo{
-							Name: directInputs[i].Branch,
-							Head: commitInfo.Commit,
-						})
-					}
-					select {
-					case <-ctx.Done():
+					commitSetsMutex.Unlock()
+				}
+			}()
+		}
+		if input.Cron != nil {
+			schedule, err := cron.Parse(input.Cron.Spec)
+			// it shouldn't be possible to error here because we validate the spec in CreatePipeline
+			if err != nil {
+				cancel()
+				return nil, err
+			}
+			tstamp := &types.Timestamp{}
+			var buffer bytes.Buffer
+			if err := pachClient.GetFile(input.Cron.Repo, "master", "time", 0, 0, &buffer); err != nil && !isNotFoundErr(err) {
+				cancel()
+				return nil, err
+			} else if err != nil {
+				// File not found, this happens the first time the pipeline is run
+				tstamp = input.Cron.Start
+			} else {
+				if err := jsonpb.UnmarshalString(buffer.String(), tstamp); err != nil {
+					cancel()
+					return nil, err
+				}
+			}
+			t, err := types.TimestampFromProto(tstamp)
+			if err != nil {
+				cancel()
+				return nil, err
+			}
+			go func() {
+				for {
+					if err := func() error {
+						nextT := schedule.Next(t)
+						t = nextT
+						time.Sleep(time.Until(nextT))
+						commit, err := pachClient.StartCommit(input.Cron.Repo, "master")
+						if err != nil {
+							return err
+						}
+						timestamp, err := types.TimestampProto(nextT)
+						if err != nil {
+							return err
+						}
+						timeString, err := (&jsonpb.Marshaler{}).MarshalToString(timestamp)
+						if err != nil {
+							return err
+						}
+						if err := pachClient.DeleteFile(input.Cron.Repo, "master", "time"); err != nil {
+							return err
+						}
+						if _, err := pachClient.PutFile(input.Cron.Repo, "master", "time", strings.NewReader(timeString)); err != nil {
+							return err
+						}
+						if err := pachClient.FinishCommit(input.Cron.Repo, "master"); err != nil {
+							return err
+						}
+						commitInfo, err := pachClient.InspectCommit(commit.Repo.Name, "master")
+						if err != nil {
+							return err
+						}
+						commitSetsMutex.Lock()
```

Comment:
It is, I'll try to refactor a bit although I may need to grab you for help. That could is still a bit mystifying to me.